README.md
sparsebnUtils
=============

[![Project Status: Active The project has reached a stable, usable state
and is being actively
developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/)
[![Travis-CI Build
Status](https://travis-ci.org/itsrainingdata/sparsebnUtils.svg?branch=master)](https://travis-ci.org/itsrainingdata/sparsebnUtils)
[![](https://www.r-pkg.org:443/pkg/sparsebnUtils)](https://www.r-pkg.org:443/pkg/sparsebnUtils)
[![CRAN RStudio mirror
downloads](https://cranlogs.r-pkg.org/badges/sparsebnUtils)](https://www.r-pkg.org/pkg/sparsebnUtils)

A set of tools for representing and estimating sparse Bayesian networks
from continuous and discrete data.

Overview
--------

This package provides various S3 classes for making it easy to estimate
graphical models from data:

-   `sparsebnData` for managing experimental data with interventions.
-   `sparsebnFit` for representing the output of a DAG learning
    algorithm.
-   `sparsebnPath` for representing a solution path of estimates.

The package also provides methods for manipulating these objects and for
estimating parameters in graphical models:

-   `estimate.parameters` for directed graphs.
-   `get.precision` for undirected graphs.
-   `get.covariance` for covariance matrices.


NEWS.md
# sparsebnUtils 0.0.8

* Fixed an issue in `Matrix_to_sparse` related to a bug fix / API change in Matrix v1.3-0

# sparsebnUtils 0.0.7

* Maintenance release for R 3.6.0

# sparsebnUtils 0.0.6

## Features

* Added `nchar` argument to improve output of show.parents with long node names (#13)
* `openCytoscape` now accepts title argument -- allows multiple cytoscape networks with different names to be open at the same time

# sparsebnUtils 0.0.5

## Features

* `openCytoscape` method added for compatibility with Cytoscape app (sparsebn #4)
* `select.parameter` now works with discrete data
* Added `summary` generics for `sparsebnPath`, `sparsebnFit`, `sparsebnData`, and
`edgeList` objects
* Added `plot` generic for `sparsebnData` objects
* Added `specify.prior` method to simplify construction of black lists
* Improve output when estimate.parameters is singular (#9)

## Bug fixes

* Fixed bug which caused `permute.nodes` to throw an error when passing in a user-specified node ordering
* Correct error message for mixed data (#8)
* Update documentation for conversion methods (#11)

# sparsebnUtils 0.0.4

## Features

* Added new `random.graph` method to generate random `edgeList`s

## Notes

* Conversion to and from sparse matrices using `as.sparse` is now significantly faster and supports `Matrix` input

# sparsebnUtils 0.0.3

## Major changes

* `get.solution()` has been renamed `select()` (for consistency with `select.parameter()`)

## Features

* Added a `NEWS.md` file to track changes to the package
* `select()` now uses fuzzy matching by default
* Added new method `get.nodes()` to return node names from a `sparsebn` object
* Improved output and print method for `estimate.parameters()` when using discrete data

## Bug fixes

* Graphics layout for `plot.sparsebnPath()` now accounts for omitted null graph
* `random.dag()` now correctly accepts `FUN` for user-specification of RNG

# sparsebnUtils 0.0.2

* Initial stable release



R/s3-sparse.R
#
#  s3-sparse.R
#  sparsebnUtils
#
#  Created by Bryon Aragam (local) on 1/22/16.
#  Copyright (c) 2014-2017 Bryon Aragam. All rights reserved.
#

#------------------------------------------------------------------------------#
# sparse S3 Class for R
#------------------------------------------------------------------------------#

#
# An alternative data structure for storing sparse matrices in R using the (row, column, value)
#   format. Internally it is stored as a list with three components, each vectors, that contain
#   the rows / columns / values of the nonzero elements.
#
# Its main purpose is to serve as an intermediary between the standard R dense matrix class and the
#   internal SparseBlockMatrixR class. That is, to convert from matrix to SBM, we do
#
#       matrix -->> sparse -->> SparseBlockMatrixR
#
# In theory, this class can be used externally as a useful data structure for storing sparse matrices
#   as an alternative to the Matrix class provided by the Matrix package. Currently, however, the class
#   structure is fairly limited, so there isn't much a reason to do this.
#
#

#' sparse class
#'
#' Low-level representation of sparse matrices.
#'
#' An alternative data structure for storing sparse matrices in R using the (row, column, value)
#' format. Internally it is stored as a list with three components, each vectors, that contain
#' the rows / columns / values of the nonzero elements.
#'
#' @param x Various \code{R} objects.
#' @param ... (optional) additional arguments.
#'
#' @docType class
#' @name sparse
NULL

#------------------------------------------------------------------------------#
# is.sparse
#
#' @rdname sparse
#' @export
is.sparse <- function(x){
    inherits(x, "sparse")
} # END IS.SPARSE

#' as.sparse
#'
#' Methods for coercing other \code{R} objects to \code{\link{sparse}} objects.
#'
#' @param x A compatible \code{R} object.
#' @param index \code{"R"} or \code{"C"}, depending on whether to use R- or C-style indexing.
#' @param ... other parameters.
#'
#' @return
#' \code{\link{sparse}}
#'
#' @export
as.sparse <- function(x, index = "R", ...){
    sparse(x, index = "R", ...) # NOTE: S3 delegation is implicitly handled by the constructor here
}

#------------------------------------------------------------------------------#
# reIndexC.sparse
#  Re-indexing TO C for sparse objects
#
# #' @describeIn reIndexC C-style re-indexing for \link{sparse} objects.
#' @export
reIndexC.sparse <- function(x){
    if(x$start == 0){
        warning("This object already uses C-style indexing!")
        return(x)
    }

    x$rows <- x$rows - 1
    x$cols <- x$cols - 1
    x$start <- 0

    x
} # END REINDEXC.SPARSE

#------------------------------------------------------------------------------#
# reIndexR.sparse
#  Re-indexing TO R for sparse objects
#
# #' @describeIn reIndexC R-style re-indexing for \link{sparse} objects.
#' @export
reIndexR.sparse <- function(x){
    if(x$start == 1){
        warning("This object already uses R-style indexing!")
        return(x)
    }

    x$rows <- x$rows + 1
    x$cols <- x$cols + 1
    x$start <- 1

    x
} # END REINDEXR.SPARSE

#------------------------------------------------------------------------------#
# sparse.list
#  List constructor
#
#' @export
sparse.list <- function(x, ...){

    if( !is.list(x)){
        stop("Input must be a list!")
    }

    if( length(x) != 5 || any(names(x) != c("rows", "cols", "vals", "dim", "start")) || is.null(names(x))){
        stop("Input is not coercable to an object of type sparse, check list for the following (named) elements: rows, cols, vals, dim, start")
    }

    if( length(unique(lapply(x[1:3], length))) > 1){
        stop("rows / cols / vals elements have different sizes; should all have the same length (pp)!!")
    }

    if(length(x$dim) != 2){
        stop("dim attribute must have length 2!")
    }

    if(x$start != 0 && x$start != 1){
        stop("start attribute must be 0 (C-style) or 1 (R-style)!")
    }

    if(!is.integer(x$rows) || !is.integer(x$cols)){
        stop("rows / cols must both be integers!")
    }

    if(!is.numeric(x$vals)){
        stop("vals must be numeric!")
    }

    structure(x, class = "sparse")
} # END SPARSE.LIST

#------------------------------------------------------------------------------#
# sparse.matrix
#
#' @export
sparse.matrix <- function(x, index = "R", ...){
    matrix_to_sparse(x, index = "R", ...)
} # END SPARSE.MATRIX

#------------------------------------------------------------------------------#
# sparse.Matrix
#
#' @export
sparse.Matrix <- function(x, index = "R", ...){
    Matrix_to_sparse(x, index = "R", ...)
} # END SPARSE.MATRIX

#' @export
sparse.edgeList <- function(x, ...){
    nnode <- num.nodes(x)
    out <- list(rows = c(), cols = c(), vals = c(), dim = c(nnode, nnode), start = 1) # enforce R-style indexing since edgeLists are never passed to C++ (at least for the time being)
    for(j in seq_along(x)){
        child <- j
        parset <- x[[child]] # parent set of j
        out$rows <- c(out$rows, parset) # set parents of child
        out$cols <- c(out$cols, rep(child, length(parset))) # children
    }

    if(length(out$rows) != length(out$cols))
        stop("Error!")

    out$vals <- as.numeric(rep(NA, length(out$cols))) # edgeLists do not carry weight information

    sparse(out)
}

#------------------------------------------------------------------------------#
# as.matrix.sparse
#  Convert FROM sparse TO matrix
#
#' @export
as.matrix.sparse <- function(x, ...){

    if( !is.sparse(x)){
        stop("Input must be a sparse object!")
    }

    if(x$start == 0) x <- reIndexR(x) # if indexing starts at 0, adjust to start 1 instead

    m.dim <- x$dim
    m <- matrix(0, nrow = m.dim[1], ncol = m.dim[2])

    for(k in seq_along(x$vals)){
        m[x$rows[k], x$cols[k]] <- x$vals[k]
    }

    attributes(m)$dim <- x$dim
    # attributes(m)$dimnames <- list()
    rownames(m) <- as.character(1:nrow(m))
    colnames(m) <- as.character(1:ncol(m))

    m
} # END AS.MATRIX.SPARSE

#------------------------------------------------------------------------------#
# as.list.sparse
#  Convert FROM sparse TO list
#
#' @export
as.list.sparse <- function(x, ...){
    list(rows = x$rows, cols = x$cols, vals = x$cols, dim = x$dim, start = x$start)
} # END AS.LIST.SPARSE

#------------------------------------------------------------------------------#
# print.sparse
#  Print function for sparse objects
#  By default, format the output as a three-column matrix [cols | rows | vals] ordered by increasing columns.
#    Optionally, set pretty = FALSE to print the sparse object as a list.
#' @export
print.sparse <- function(x, pretty = TRUE, ...){
    if(pretty){
        out <- cbind(x$cols, x$rows, x$vals)
        colnames(out) <- c("cols", "rows", "vals")
        print(out)
    } else{
        print(as.list(x))
    }

} # END PRINT.SPARSE

#------------------------------------------------------------------------------#
# is.zero.sparse
#  Check to see if a sparse object represents the zero matrix
#
#' @export
is.zero.sparse <- function(x){
    check_if_zero <- (length(x$rows) == 0)

    check_if_zero
} # END IS.ZERO.SPARSE

#------------------------------------------------------------------------------#
# num.nodes.sparse
#
#' @export
num.nodes.sparse <- function(x){
    x$dim[2]
} # NUM.NODES.SPARSE

#------------------------------------------------------------------------------#
# num.edges.sparse
#
#' @export
num.edges.sparse <- function(x){
    ### What to do about this special case...
    # length(x$rows) # Ignores potentially very small edge weights which may be zero

    .num_edges(x)
} # NUM.EDGES.SPARSE

#------------------------------------------------------------------------------#
# t.sparse
#  Take implicit transpose by swapping rows <-> cols
#
#' @export
t.sparse <- function(x){
    temp <- x$rows
    x$rows <- x$cols
    x$cols <- temp

    x
}

#------------------------------------------------------------------------------#
# .num_edges
# Internal function for returning the number of edges in a sparse object
#
.num_edges <- function(x, threshold = FALSE){
    stopifnot(is.sparse(x))

    if(!threshold){
        length(x$rows)
    } else{
        ### Testing only for now
        if(length(which(abs(x$vals) > zero_threshold())) != length(x$rows)){
            stop("Error in .num_edges.sparse! Please check source code.")
        }

        length(which(abs(x$vals) > zero_threshold()))
    }

} # END .NUM_EDGES

matrix_to_sparse <- function(x, index = "R", ...){
    stopifnot(check_if_matrix(x))

    if( nrow(x) != ncol(x)) stop("Input matrix must be square!") # 2-7-15: Why does it need to be square?

    if(index != "R" && index != "C") stop("Invalid entry for index parameter: Must be either 'R' or 'C'!")

    pp <- nrow(x)

    # t1 <- proc.time()[3] ################################################
    nnz <- Matrix::which(is.na(x) | (abs(x) > zero_threshold()), arr.ind = TRUE)
    # t2 <- proc.time()[3] ################################################
    # cat(sprintf("as.sparse nnz via which(...): %f\n", t2-t1))

    ### This is a weird hack that is surprisingly fast:
    ###  Note that the output of which is in the same order as as.vector,
    ###  in the sense that the first element in as.vector(x[x != 0])
    ###  corresponds exactly to the first row of nnz (which is the output
    ###  of which).
    ###
    ### This is _substantially_ faster than either Matrix::Matrix
    ###  and an Rcpp implementation of the loop that was previously used.
    vals <- as.vector(x[x != 0])
    rows <- nnz[, 1]
    cols <- nnz[, 2]

    sp <- sparse.list(list(rows = as.integer(rows), cols = as.integer(cols), vals = as.numeric(vals),
                           dim = c(pp, pp),
                           start = 1
                           )
                      )

    if(index == "R"){
        suppressWarnings(reIndexR(sp))
    } else{
        suppressWarnings(reIndexC(sp))
    }
} # END MATRIX_TO_SPARSE

Matrix_to_sparse <- function(x, index = "R", ...){
    stopifnot(check_if_matrix(x))

    if( nrow(x) != ncol(x)) stop("Input matrix must be square!") # 2-7-15: Why does it need to be square?

    if(index != "R" && index != "C") stop("Invalid entry for index parameter: Must be either 'R' or 'C'!")

    pp <- nrow(x)

    ### If already a Matrix object, just pull out the corresponding slots
    #
    # Tue Jan 12 2021: As of Matrix  1.3-0 (2020-12-15 r3351), this code breaks
    #                   due to a bugfix in the Matrix package. The issue
    #                   according to the Matrix maintainers is that "you call
    #                   a function e.g.
    #                      as(M, "dgTMatrix")  or chol()  or ..
    #                   on such a Matrix but that method does not work, or not
    #                   work as assumed, now that Matrix() does return a
    #                   "ddiMatrix" when it can & unless you explicitly say
    #                   ' doDiag=FALSE '.
    #
    #                   The fix to just say as(x, "TsparseMatrix") instead of
    #                   as(x, "dgTMatrix") below.
    #
    Tx <- as(x, "TsparseMatrix") # needs to be in triplet form, see ?dgCMatrix vs ?dgTMatrix
    vals <- Tx@x
    rows <- Tx@i
    cols <- Tx@j

    sp <- sparse.list(list(rows = as.integer(rows), cols = as.integer(cols), vals = as.numeric(vals),
                           dim = c(pp, pp),
                           start = 0 # NOTE: dgTMatrix uses C-style indexing, so this is different from matrix_to_sparse!
                           )
                      )

    if(index == "R"){
        suppressWarnings(reIndexR(sp))
    } else{
        suppressWarnings(reIndexC(sp))
    }
} # END MATRIX_TO_SPARSE


R/sparsebnUtils-lambdas.R
#
#  sparsebnUtils-lambdas.R
#  sparsebnUtils
#
#  Created by Bryon Aragam (local) on 1/22/16.
#  Copyright (c) 2014-2017 Bryon Aragam. All rights reserved.
#

#
# PACKAGE SPARSEBNUTILS: Lambdas
#
#   CONTENTS:
#     generate.lambdas
#     gen_lambdas
#

#' generate.lambdas
#'
#' Convenience function for creating a grid of lambdas.
#'
#' See Section 5.3 of \href{https://jmlr.org/papers/v16/aragam15a.html}{Aragam and Zhou (2015)}
#' for a discussion of regularization paths (also, solution paths).
#'
#' @param lambda.max Maximum value of lambda; in terms of the algorithm this is the initial value
#'                   of the regularization parameter in the solution path.
#' @param lambdas.ratio Ratio between the maximum lambda value and the minimum lambda value in the solution
#'                      path.
#' @param lambdas.length Number of values to include.
#' @param scale Which scale to use: Either \code{"linear"} or \code{"log"}.
#'
#' @export
generate.lambdas <- function(lambda.max,
                             lambdas.ratio = 1e-3,
                             lambdas.length = 50,
                             scale = "linear"
){
    lambda.min <- lambdas.ratio * lambda.max
    gen_lambdas(lambda.max = lambda.max,
                lambda.min = lambda.min,
                lambdas.length = lambdas.length,
                scale = scale)
} # END GENERATE.LAMBDAS

# gen_lambdas
#  Internal implementation of generate.lambdas
gen_lambdas <- function(lambda.max,
                        lambda.min,
                        lambdas.length = 50,
                        scale = "linear"
){
#     lambda.max <- max(sqrt(nn))
#     lambda.min <- lambdas.ratio * lambda.max

    if(scale == "linear"){
        lambdas <- seq(lambda.max, lambda.min, length.out = lambdas.length)
    } else if(scale == "log"){
        lambdas.ratio <- lambda.min / lambda.max
        lambdas <- exp(seq(log(lambda.max), log(lambda.min), log(lambdas.ratio)/(lambdas.length-1)))
    } else{
        stop("Invalid input for scale argument! Must be either 'log' or 'linear'.")
    }

    lambdas
} # END GEN_LAMBDAS


R/sparsebnUtils-cytoscape.R
#
#  sparsebnUtils-cytoscape.R
#  sparsebnUtils
#
#  Created by Bryon Aragam (local) on 9/1/17.
#  Copyright (c) 2014-2017 Bryon Aragam. All rights reserved.
#

#
# PACKAGE SPARSEBNUTILS: Compatibility with Cytoscape app
#
#   CONTENTS:
#       openCytoscape
#       showCytoscape
#

#' Display graphs in Cytoscape
#'
#' NOTE: This method is currently experimental and under development!
#'
#' Displays the selected graph in the Cytoscape application. Note that this
#' requires that Cytoscape is installed on the user's system, and that the
#' RCy3 package is installed and properly configured. Cytoscape can be
#' downloaded at \href{https://cytoscape.org/}{https://cytoscape.org/}.
#'
#' @param x A \code{\link{sparsebnFit}} object or other graph object.
#' @param title A character string, this is the name you will see on the Cytoscape
#'              network window. Multiple windows with the same name are not
#'              permitted. See \code{\link[RCy3]{createNetworkFromGraph}} for more
#'              details.
#' @param ... Other arguments to \code{\link[RCy3]{createNetworkFromGraph}}.
#'
#' @export
openCytoscape  <- function(x, title, ...){
    ### This function requires the 'igraph' package to be installed
    if (!requireNamespace("igraph", quietly = TRUE)) {
        stop("This method requires the igraph package; please install it first via install.packages('igraph').", call. = FALSE)
    }

    UseMethod("openCytoscape", x)
} # END OPENCYTOSCAPE

#' @export
openCytoscape.sparsebnPath <- function(x, title, ...){
    stop("Currently, this method only works on individual networks, and is not yet implemented for solution paths. Please select one network from the path instead (see ?select, ?select.parameter). Stay tuned, there will be an update for solution paths soon!")
} # END OPENCYTOSCAPE.SPARSEBNPATH

#' @export
openCytoscape.sparsebnFit <- function(x, title, ...){
    ### Convert to graphNEL object, which is needed for RCy3
    # graph <- to_igraph(x)
    # graph <- igraph::as_graphnel(graph$edges)
    graph <- to_graphNEL(x$edges)

    ### NOTE: Need to remove metadata based on current implementation of to_igraph
    ### TODO: Update this to handle the metadata more gracefully
    graph@edgeData@defaults <- list()
    graph@edgeData@data <- list()

    showCytoscape(graph, title, ...)
} # END OPENCYTOSCAPE.SPARSEBNFIT

#' @export
openCytoscape.default <- function(x, title, ...){
    ### Convert to graphNEL object, which is needed for RCy3
    # graph <- to_igraph(x)
    # graph <- igraph::as_graphnel(graph)
    graph <- to_graphNEL(x)

    ### NOTE: Need to remove metadata based on current implementation of to_igraph
    ### TODO: Update this to handle the metadata more gracefully
    graph@edgeData@defaults <- list()
    graph@edgeData@data <- list()

    showCytoscape(graph, title, ...)
} # END OPENCYTOSCAPE.DEFAULT

### Utility function to handle all RCy3-related functions including
### opening cytoscape and displaying the graph
###  Expects graphNEL input; openCytoscape handles any and all conversions
showCytoscape <- function(graph, title, ...){
    ### This function requires the 'graph' package to be installed
    if (!requireNamespace("graph", quietly = TRUE)) {
        stop("This method requires the graph package; please install it first via BioConductor.", call. = FALSE)
    }

    ### This function requires the 'graph' package to be installed
    if (!requireNamespace("RCy3", quietly = TRUE)) {
        stop("This method requires the RCy3 package to connect to Cytoscape; please install it first via BioConductor.", call. = FALSE)
    }

    stopifnot(inherits(graph, "graphNEL"))

    ### Open cytoscape window and show graph
    # cytowin <- RCy3::CytoscapeWindow(title = title,
    #                                  graph = graph,
    #                                  ...)
    cytowin <- RCy3::createNetworkFromGraph(title = title,
                                            graph = graph,
                                            ...)

    # RCy3::displayGraph(cytowin)
    RCy3::layoutNetwork(network = cytowin, layout.name = "hierarchical")
} # END SHOWCYTOSCAPE


R/s3-sparsebnPath.R
#
#  s3-sparsebnPath.R
#  sparsebnUtils
#
#  Created by Bryon Aragam (local) on 1/22/16.
#  Copyright (c) 2014-2017 Bryon Aragam. All rights reserved.
#

#------------------------------------------------------------------------------#
# sparsebnPath S3 Class for R
#------------------------------------------------------------------------------#

#
# sparsebnPath S3 class skeleton
#
# Data
# * <wrapper for a list>
#
# Methods
# * is.sparsebnPath
# * sparsebnPath.list
# * print.sparsebnPath
# * as.list.sparsebnPath
# * num.nodes.sparsebnPath
# * num.edges.sparsebnPath
# * num.samples.sparsebnPath
# * get.lambdas.sparsebnPath
# * get.adjacency.matrix.sparsebnPath
#

#' sparsebnPath class
#'
#' Convenience wrapper class for solution paths of DAG learning algorithms: This class represents an entire
#' solution path of an algorithm. Its components are of type \code{\link{sparsebnFit}}. Also inherits
#' from \code{\link{list}}.
#'
#' Each value of lambda in the (discrete) solution path corresponds to a single DAG estimate (see \href{https://jmlr.org/papers/v16/aragam15a.html}{Aragam and Zhou (2015)} for details).
#' Internally, this estimate is represented by a \code{\link{sparsebnFit}} object. The full solution
#' path is then represented as a \code{\link{list}} of \code{\link{sparsebnFit}} objects: This class is essentially a wrapper for this list.
#'
#' Most methods for \code{\link{sparsebnPath}} objects simply apply \code{\link{lapply}} to the
#' object in question. The exceptions to this rule apply when the output will always be the same
#' for every component; e.g. \code{\link{num.nodes}} and \code{\link{num.samples}}.
#'
#' @param x A \code{list} or an object of type \code{sparsebnPath}. Should only be used internally.
#' @param ... (optional) additional arguments.
#'
#' @section Methods:
#' \code{\link{get.adjacency.matrix}}, \code{\link{get.lambdas}},
#' \code{\link{num.nodes}}, \code{\link{num.edges}}, \code{\link{num.samples}}
#'
#' @examples
#'
#' \dontrun{
#' ### Learn the cytometry network
#' library(sparsebn)
#' data(cytometryContinuous) # from the sparsebn package
#' cyto.data <- sparsebnData(cytometryContinuous[["data"]], type = "continuous")
#' cyto.learn <- estimate.dag(cyto.data)
#'
#' ### Inspect the output
#' class(cyto.learn)
#' print(cyto.learn)
#' plot(cyto.learn)
#' }
#'
#' @docType class
#' @name sparsebnPath
NULL

#' @rdname sparsebnPath
#' @export
is.sparsebnPath <- function(x){
    inherits(x, "sparsebnPath")
} # END IS.sparsebnPath

# sparsebnPath constructor
#' @export
sparsebnPath.list <- function(x){
    if(!check_list_class(x, "sparsebnFit")){
        stop("Some component is not of type sparsebnPath -- sparsebnPath objects must consist of sparsebnFit components only.")
    }

    ### Note that we still allow these objects to inherit from the base list class
    structure(x, class = c("sparsebnPath", "list"))
} # END sparsebnPath.LIST

.str_sparsebnPath <- function(x){
    sbp.out <- ""
    sbp.out <- paste0(sbp.out,
                      "sparsebn Solution Path\n",
                      " ", num.nodes(x), " nodes\n",
                      " ", num.samples(x), " observations\n",
                      " ", length(x), " estimates for lambda in [", round(min(get.lambdas(x)), 4), ", ", round(max(get.lambdas(x)), 4), "]\n",
                      " ", "Number of edges per solution: ", paste(num.edges(x), collapse = "-"), "\n"
                )

    sbp.out
} # END .STR_SPARSEBNPATH

#' @param verbose If \code{TRUE}, then each estimate in the solution path is printed separately. Do not use for
#'        large graphs or large solution paths. (default = \code{FALSE})
#'
#' @rdname sparsebnPath
#' @method print sparsebnPath
#' @export
print.sparsebnPath <- function(x, verbose = FALSE, ...){
    if(verbose){
        print.default(x) # default generic reverts to list => separate calls to print.sparsebnFit for each component
    } else{
        cat(.str_sparsebnPath(x))
    }
} # END PRINT.SPARSEBNPATH

#' @param object an object of type \code{sparsebnPath}
#'
#' @rdname sparsebnPath
#' @method summary sparsebnPath
#' @export
summary.sparsebnPath <- function(object, ...){
    ### Print usual sparsebnPath output
    cat(.str_sparsebnPath(object))
    cat("\n")

    ### Add summary for each lambda
    lambdas <- get.lambdas(object)
    nedges <- num.edges(object)
    print(data.frame(lambda = lambdas, nedge = nedges))

    # sbp.out <- sprintf("%10.4f %5d", round(lambdas, 4), nedges)
    # cat(sbp.out, sep = "\n")
} # END SUMMARY.SPARSEBNPATH

#' @param labels \code{TRUE} or \code{FALSE}. Whether or not to print out
#' labels with summary information for each plot in the solution path.
#'
#' @rdname sparsebnPath
#' @method plot sparsebnPath
#' @export
plot.sparsebnPath <- function(x, labels = FALSE, ...){
    ### UPDATE 7/28/17: What are the issues? Seems fine.
    ### Issues when plotting null DAG, so remove it
    # x <- x[-1] # Do this BEFORE setting the grid layout below!

    ### Set plotting parameters (Don't use no.readonly = TRUE! See https://stat.ethz.ch/pipermail/r-help/2007-July/136770.html)
    par.default <- par()[c("mfrow", "mai")] # Only re-set what we change here
    par(mfrow = n2mfrow(length(x)),         # Automatically choose a sensible grid to use
        mai = c(0, 0, 0.1, 0)                      # Need to reset margins (why??? graph packages seem to handle this oddly)
        )

    tryCatch({
        # lapply(x, plot)
        for(fit in x){
            plot(fit, ...)
            if(labels){
                # title(sprintf("lambda = %4.2f, # of edges = %d", fit$lambda, fit$nedge), cex.main = 0.85)
                title(bquote(lambda*" = "*.(round(fit$lambda, 2))*" ("*.(fit$nedge)*" edges)"), cex.main = 0.85)
            }
        }
    }, error = function(c){
        dev.off()
        stop(c)
    })

    par(par.default) # restore user's original settings
}

#' @export
as.list.sparsebnPath <- function(x, ...){
    class(x) <- "list"
    x
} # END AS.LIST.SPARSEBNPATH

#' @describeIn num.nodes Extracts the number of nodes of \link{sparsebnPath} object.
#' @export
num.nodes.sparsebnPath <- function(x){
    unique(unlist(lapply(x, function(z) z$pp)))
} # END NUM.NODES.SPARSEBNPATH

#' @describeIn num.edges Extracts the number of edges of \link{sparsebnPath} object.
#' @export
num.edges.sparsebnPath <- function(x){
    ### unique(.) not needed since different estimates should have different # of edges
    unlist(lapply(x, function(z) z$nedge))
} # END NUM.EDGES.SPARSEBNPATH

#' @describeIn num.samples Extracts the number of samples of \link{sparsebnPath} object.
#' @export
num.samples.sparsebnPath <- function(x){
    unique(unlist(lapply(x, function(z) z$nn)))
} # END NUM.SAMPLES.SPARSEBNPATH

#' Extract regularization path from solution path
#'
#' @describeIn get.lambdas Returns a vector of lambda values defining the solution path of a \code{\link{sparsebnPath}} object.
#' @export
get.lambdas.sparsebnPath <- function(x){
    lambdas <- unlist(lapply(x, function(z){ z$lambda}))
    names(lambdas) <- NULL

    lambdas
} # END GET.LAMBDAS.SPARSEBNPATH

#' Extract node names from solution path
#'
#' @describeIn get.nodes Returns the node names from a \code{\link{sparsebnPath}} object.
#' @export
get.nodes.sparsebnPath <- function(x){
    x[[1]]$nodes
} # END GET.NODES.SPARSEBNPATH

#' @describeIn get.adjacency.matrix Retrieves all \code{edges} slots in the solution path, converts to an adjacency matrix, and returns as a list
#' @export
get.adjacency.matrix.sparsebnPath <- function(x){
    lapply(x, get.adjacency.matrix)
} # END GET.ADJACENCY.MATRIX.sparsebnPath

#' @export
"[.sparsebnPath" <- function(x, i){
    ### Needed to ensure that naive subsetting of path objects as a list
    ###  still returns a path object
    sparsebnPath(as.list(x)[i])
}


#' @export
to_edgeList.sparsebnPath <- function(x){
    sparsebnPath(lapply(x, to_edgeList))
}

